### PR TITLE
ATO-827: Authorize endpoint

### PR DESCRIPTION
**Pull request by Ryan-Andrews99**
## What

- The simulator now has a `/authorize` endpoint. 
- This responds with a redirect to the valid redirect_uri with a fixed authcode 
- The code and parameters are stored in the AuthCode param store
- The `/authorize` request is parsed and validated against the client config and valid OIDC values
- 

## How to review: 
- Code review commit by commit 
- Look at the tests in `authorise/tests/authorise-get-controller.test.ts` to see the outward API responses for a range of valid and invalid requests

### CODE REVIEW

File: src/parse/parse-auth-request-query-params.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+  }
+
+  if (!queryParams.response_type) {
+    throw new ParseRequestError(
+      "Invalid Request: Invalid response_type parameter",
+      queryParams.redirect_uri,
+      queryParams.client_id
+    );
+  }
+
```
Comment (CarlyG55) [inline src/parse/parse-auth-request-query-params.ts:68]: I think there's a prompt check missing here? oh wait it's below, but again think it should be higher up?

File: src/parse/parse-auth-request-query-params.ts [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    throw new MissingParameterError(
+      "Invalid Request: Missing client_id parameter"
+    );
+  }
+
+  if (!queryParams.redirect_uri) {
+    throw new MissingParameterError(
+      "Invalid Request: Invalid redirect_uri parameter"
+    );
+  }
```
Comment (CarlyG55) [inline src/parse/parse-auth-request-query-params.ts:53]: I think this comes later, just before scope?

File: src/parse/parse-auth-request-query-params.ts [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+  if (!queryParams.scope) {
+    throw new ParseRequestError(
+      "Invalid Request: Missing scope parameter",
+      queryParams.redirect_uri,
+      queryParams.client_id
+    );
+  }
+
```
Comment (CarlyG55) [inline src/parse/parse-auth-request-query-params.ts:76]: response type is another example of something we don't explicitly ask for, but should check we give the same response for

File: src/parse/parse-auth-request-query-params.ts [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+  const scopes = queryParams.scope.split(" ");
+
+  if (!scopes.includes("openid")) {
+    throw new ParseRequestError(
+      "Invalid Request: The scope must include an openid value",
+      queryParams.redirect_uri,
+      queryParams.client_id
+    );
+  }
```
Comment (CarlyG55) [inline src/parse/parse-auth-request-query-params.ts:96]: think this should be before the client id/redirect uri checks just above?

File: src/parse/parse-auth-request-query-params.ts (hunk 3 above)
Comment (CarlyG55) [inline src/parse/parse-auth-request-query-params.ts:76]: also missing ui_locales checks (maybe? do we need them?)

File: src/parse/parse-auth-request-query-params.ts [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+  const parsedClaims: string[] = [];
```
Comment (CarlyG55) [inline src/parse/parse-auth-request-query-params.ts:98]: picky but can we declare this next to where it's used? makes it clearer it's not used between the two places

File: src/parse/parse-auth-request-query-params.ts [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+  if (queryParams.claims && !isValidJSON(queryParams.claims)) {
+    throw new ParseRequestError(
+      "Invalid Request: Invalid claims parameter",
+      queryParams.redirect_uri,
+      queryParams.client_id
+    );
+  }
+
+  const userInfoClaims = JSON.parse(queryParams.claims ?? "{}").userinfo;
```
Comment (CarlyG55) [inline src/parse/parse-auth-request-query-params.ts:108]: are we parsing claims twice..? I thought we do it in the `isClaimsValid`

File: src/components/authorise/authorise-get-controller.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+    config.addToAuthCodeRequestParamsStore(authCode, {
+      claims: parsedAuthRequestParams.claims,
+      nonce: parsedAuthRequestParams.nonce,
+      redirectUri: parsedAuthRequestParams.redirect_uri,
+      scopes: parsedAuthRequestParams.scope,
+    });
+
+    res.redirect(
+      `${parsedAuthRequestParams.redirect_uri}?code=${authCode}&state=${parsedAuthRequestParams.state}&nonce=${parsedAuthRequestParams.nonce}`
```
Comment (CarlyG55) [inline src/components/authorise/authorise-get-controller.ts:27]: I'm pretty sure we don't return the nonce in this request?

File: src/components/authorise/authorise-get-controller.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    } else {
+      res.redirect(
+        `${error.redirect_uri}?error=invalid_request&error_description=${encodeURIComponent("Invalid Request")}`
+      );
+    }
+  } else if (error instanceof BadRequestError) {
+    res.status(400).send("Invalid Request");
+  } else {
+    logger.error("Unknown error occurred: " + (error as Error).message);
+    res.status(502).json({
```
Comment (CarlyG55) [inline src/components/authorise/authorise-get-controller.ts:70]: should this not be 500?

File: src/components/authorise/authorise-get-controller.ts (hunk 7 above)
Comment (Ryan-Andrews99) [reply src/components/authorise/authorise-get-controller.ts:27]: Yeah we don't, I have no idea why I added this 🤦

File: src/validators/vtr-validator.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+      "P2 identity confidence must require at least Cl.Cm credential trust"
+    );
+  }
+
+  return parsedVtr;
+};
+
+const isCredentialTrust = (
+  credentialTrust: string
+): credentialTrust is "Cl" | "Cl.Cm" => credentialTrust.startsWith("C");
```
[the remark this answers is not in the document]
Comment (Ryan-Andrews99) [reply src/validators/vtr-validator.ts]: Fixed!

File: src/validators/vtr-validator.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+const parseSingleVtr = (singleVtr: string): VectorOfTrust => {
+  const splitVtrValues = singleVtr.split(".");
+
+  const credentialTrustValue = splitVtrValues
+    .filter(isCredentialTrust)
+    .sort()
+    .join(".");
+
+  const levelsOfConfidence = splitVtrValues.filter(isVectorOfTrust);
```
[the remark this answers is not in the document]
Comment (Ryan-Andrews99) [reply src/validators/vtr-validator.ts]: Changed this

File: src/validators/vtr-validator.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  if (!VALID_CREDENTIAL_TRUST_VALUES.includes(credentialTrustValue)) {
+    throw new Error("Invalid Credential Trust Value provided");
+  }
+
+  if (
+    levelsOfConfidence[0] &&
+    !VALID_LOC_VALUES.includes(levelsOfConfidence[0])
+  ) {
```
[the remark this answers is not in the document]
Comment (Ryan-Andrews99) [reply src/validators/vtr-validator.ts]: Yep, looking at the code this was slightly incorrect, have fixed!

File: src/validators/vtr-validator.ts
[the remark this answers is not in the document]
Comment (Ryan-Andrews99) [reply src/validators/vtr-validator.ts:1]: Yep have given it another go!

File: src/errors/duplicate-parameter-error.ts
[the remark this answers is not in the document]
Comment (Ryan-Andrews99) [reply src/errors/duplicate-parameter-error.ts:1]: Have added some middleware to fix this

File: src/parse/parse-auth-request-query-params.ts (hunk 4 above)
Comment (Ryan-Andrews99) [reply src/parse/parse-auth-request-query-params.ts]: Have moved!

File: src/parse/parse-auth-request-query-params.ts (hunk 3 above)
Comment (Ryan-Andrews99) [reply src/parse/parse-auth-request-query-params.ts:72]: Have added a small check that logs a warning if something unsupported is requested, but wasn't sure on parsing and checking the lang tags. I'm happy to leave this out

File: src/parse/parse-auth-request-query-params.ts (hunk 5 above)
Comment (Ryan-Andrews99) [reply src/parse/parse-auth-request-query-params.ts]: Have amended this more generally so this is resolved!

File: src/parse/parse-auth-request-query-params.ts (hunk 6 above)
Comment (Ryan-Andrews99) [reply src/parse/parse-auth-request-query-params.ts]: Have fixed this!

File: src/components/authorise/authorise-get-controller.ts (hunk 7 above)
Comment (Ryan-Andrews99) [reply src/components/authorise/authorise-get-controller.ts]: Fixed

File: src/components/authorise/authorise-get-controller.ts (hunk 8 above)
Comment (Ryan-Andrews99) [reply src/components/authorise/authorise-get-controller.ts]: Fixed

File: src/parse/parse-auth-request-query-params.ts [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+    );
+  }
+
+  if (!queryParams.response_type) {
+    throw new ParseRequestError(
+      "Invalid Request: Invalid response_type parameter",
+      queryParams.redirect_uri,
+      queryParams.client_id
+    );
+  }
```
[the remark this answers is not in the document]
Comment (Ryan-Andrews99) [reply src/parse/parse-auth-request-query-params.ts]: Fixed

File: src/validators/vtr-validator.ts
Comment (ethanmills) [inline src/validators/vtr-validator.ts:1]: I might have a go at refactoring this after it's gone in. It's a gnarly bit of logic and I wonder if we can express it differently somehow

File: tests/integration/app.test.ts [hunk 13]
```diff
@@ -3,7 +3,7 @@ import request from "supertest";
 
 describe("Integration: Healthcheck", () => {
   test("Healthcheck endpoint returns 200", async () => {
-    const app = await createApp();
+    const app = createApp();
```
Comment (ethanmills) [inline tests/integration/app.test.ts:6]: Why is this needed, just out of curiosity?

**Review by ethanmills — DISMISSED** [review]

File: src/validators/vtr-validator.ts [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+    const identityVectors = parsedVtrSet
+      .filter((vector) => !!vector.levelOfConfidence)
+      .map((vector) => vector.levelOfConfidence) as ("P0" | "P1" | "P2")[];
+
+    if (
+      identityVectors.length !== 0 &&
+      identityVectors.length < parsedVtrSet.length
+    ) {
+      logger.error("VTR cannot contain both identity and non-identity vectors");
```
[the remark this answers is not in the document]
Comment (Ryan-Andrews99) [reply src/validators/vtr-validator.ts]: It is, pushed some changes to ensure P0 counts as a non-identity vector here, which matches our logic

File: src/validators/vtr-validator.ts
Comment (Ryan-Andrews99) [reply src/validators/vtr-validator.ts:1]: Have done a little refactoring to make it read a bit better, but yeah remains kinda gnarly

File: tests/integration/app.test.ts (hunk 13 above)
Comment (Ryan-Andrews99) [reply tests/integration/app.test.ts:6]: I think the test was setup as an example when the repo was created

File: src/constants.ts [hunk 15]
```diff
@@ -1,3 +1,33 @@
...
+  "https://vocab.account.gov.uk/v1/drivingPermit",
+  "https://vocab.account.gov.uk/v1/socialSecurityRecord",
+  "https://vocab.account.gov.uk/v1/coreIdentityJWT",
+  "https://vocab.account.gov.uk/v1/returnCode",
+  "https://vocab.account.gov.uk/v1/inheritedIdentityJWT",
+];
+
+export const VALID_CREDENTIAL_TRUST_VALUES = ["Cl", "Cl.Cm"];
+
+export const VALID_LOC_VALUES = ["P0", "P2"];
```
[the remark this answers is not in the document]
Comment (Ryan-Andrews99) [reply src/constants.ts]: Done!

**Review by ethanmills — APPROVED** [review]
